### PR TITLE
Add initial implementation of plugin registry

### PR DESCRIPTION
**Pull request by davidbeckingsale**
This will allow us to link in plugins that can act before and after kernel launches. One major benefit will be to fix the backward RAJA-CHAI dependency, meaning we can move the CHAI support code to the CHAI repo

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
We probably need to decide if we want "named" kernels for that 😎

**Comment by trws** [comment]
Name would be a nice thing to have, but even just having it be by __PRETTY_FUNCTION__ or something would be better than what we have now.

**Comment by DavidPoliakoff** [comment]
I think I have an excuse to write that plugin, if we get to that point. I'm actually working literally right now on Annotation policies in Caliper that could play nicely with this...

Let me get that PR up, then we should talk

**Comment by DavidPoliakoff** [comment]
Hey folks, in [this PR](https://github.com/LLNL/Caliper/pull/108) I set up the ability for applications to have "tagged" annotations which are statically enabled or disabled by policy. I'm imagining the ability to set up a plugin using this, so people could decide whether or not to have an instrumented RAJA quite simply. @mplegendre and I will talk offline about whether we want to target this. Very cool stuff.

**Comment by DavidPoliakoff** [comment]
@davidbeckingsale : one thing I'm worried about is how this will work with our copy construction mechanisms in the debugging work. We should talk offline about it, but one thing that would be nice would be a way to tell whether a copy constructor was responding to [this guy](https://github.com/LLNL/RAJA/blob/develop/include/RAJA/pattern/forall.hpp#L213) or something else. I almost wonder whether we want to move the preLaunch trigger into trigger_updates_before, but that seems a little gruesome as well.

**Comment by davidbeckingsale** [comment]
@trws @ajkunen - can you review this?

File: include/RAJA/internal/get_platform.hpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#ifndef RAJA_get_platform_HPP
+#define RAJA_get_platform_HPP
+
+#include "RAJA/internal/ForallNPolicy.hpp"
+#include "RAJA/internal/LegacyCompatibility.hpp"
```
Comment (trws) [inline include/RAJA/internal/get_platform.hpp]: Missing policybase include here for Platform enum?

**Comment by ajkunen** [comment]
this looks good.  Adding the hooks into RAJA::kernel, and completely removing the CHAI-specific stuff would be wonderful too.

**Comment by davidbeckingsale** [comment]
@ajkunen - yup, I'll do that and fix the includes as per @trws

**Comment by davidbeckingsale** [comment]
@ajkunen - what would be your preferred way to deal with the ManagedArrayView? Move to CHAI?

**Comment by ajkunen** [comment]
yeah, I think that would belong in CHAI

**Comment by davidbeckingsale** [comment]
Think this is ready to go :)

**Comment by trws** [comment]
Travis appears cranky on icc, is this a known issue?

**Comment by davidbeckingsale** [comment]
I've not seen that before 😞

**Comment by davidbeckingsale** [comment]
@trws - do you think the Intel failure is because Intel is being weird, or because there is a problem with our code? It doesn't seem to be getting the correct specialization for one of the type-traits

**Comment by davidbeckingsale** [comment]
Good point - I'll check it with the current Intel version.

**Comment by davidbeckingsale** [comment]
@trws - Intel 18 is building fine.

**Comment by davidbeckingsale** [comment]
I can reproduce the breakage with 16 and 17 locally.

**Comment by rhornung67** [comment]
@davidbeckingsale you may want to ping RAJA+CHAI app users (usual suspects) to see what version of Intel compiler they are using....

**Comment by trws** [comment]
That would definitely be good.  I pinged @jeffhammond about the license so we can add 17 or 18 to CI, but if 17 doesn't work we should probably find a workaround for the issue and gate it.  If it's something that might come up elsewhere we could add some machinery to CAMP to abstract it for that matter.

**Comment by jeffhammond** [comment]
Please use [Intel 19 beta](https://software.intel.com/en-us/articles/intel-parallel-studio-xe-2019-beta) until I resolve this.

**Comment by davidbeckingsale** [comment]
We need to support 17+

@trws  - the code that's breaking I don't think is any different from the previous code that worked with the old CHAI integration. Could it be something to do with the order or location in which the specializations are declared?

**Comment by trws** [comment]
It looks like it’s an issue with what is and what isn’t treated as 
sfinae.  Adding a layer of indirection to protect it from attempting to 
instantiate the indexset execpolicy with only one argument might work, 
or wrapping it in a decltype or… I’m not sure, it’s the kind of 
thing you’ll likely have to just play with alternate ways to express 
that requirement until something works.

File: include/RAJA/pattern/forall.hpp [hunk 2]
```diff
@@ -428,15 +452,20 @@ forall_Icount(ExecutionPolicy&& p,
               LoopBody&& loop_body)
 {
 
-  detail::setChaiExecutionSpace<ExecutionPolicy>();
+  util::PluginContext p_context;
+  p_context.platform = detail::get_platform<ExecutionPolicy>::value;
```
[the remark this answers is not in the document]
Comment (DavidPoliakoff) [reply include/RAJA/pattern/forall.hpp]: @trws happened upon this issue in an email. We have a make_context guy [here](https://github.com/kokkos/kokkos/blob/master/core/src/impl/Kokkos_AnalyzePolicy.hpp). Another thing to know is that it's not really clear whether the context should be `CUDA` or `CUDADevice<DeviceNum>`.

File: include/RAJA/util/PluginContext.hpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace RAJA {
+namespace util {
+
+struct PluginContext {
+  Platform platform;
+};
+
+template<typename Policy>
+PluginContext make_context()
```
[the remark this answers is not in the document]
Comment (trws) [reply include/RAJA/util/PluginContext.hpp]: Essentially so yeah, that way hopefully we wont have to poke quite so many places when we reconcile this with devices.  For now I don't know that we need a name, but with this we could add one by making it a region that gets pulled out in this function as a first cut.  I'm happy with this. 👍

**Comment by davidbeckingsale** [comment]
This is finally ready (!!!!), as is the corresponding CHAI PR.

Could I get some final reviews? @trws @ajkunen @rhornung67

**Comment by rhornung67** [comment]
@davidbeckingsale I'll push a change to fix the build failure.

**Comment by rhornung67** [comment]
@davidbeckingsale Build issues are fixed, but coverage check is failing. Can you take a look at that?

**Pull request merged**